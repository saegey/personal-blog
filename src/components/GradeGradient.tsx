import { useMemo } from 'react'

interface GradeGradient {
  data: Array<{
    color: string
    distance: number
  }>
  xMax: number
}

const GradeGradient = ({ data, xMax }: GradeGradient) => {
  const gradients = useMemo(() => {
    return data.map((d, i) => {
      if (
        data.length > i + 1 &&
        i > 0 &&
        d.color === data[i - 1].color &&
        d.color === data[i + 1].color
      ) {
        return
      }
      const gradient = (
        <stop
          offset={Number((d.distance / xMax).toFixed(10))}
          stopColor={d.color}
          stopOpacity={1}
          key={`elevationGrade-${i}`}
        />
      )
      return gradient
    })
  }, [data])

  return gradients
}

export default GradeGradient
