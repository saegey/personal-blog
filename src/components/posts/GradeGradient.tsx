import React from "react";

interface GradeGradientProps {
	data: Array<{
		color: string;
		distance: number;
	}>;
	xMax: number;
}

const GradeGradient = ({ data, xMax }: GradeGradientProps): JSX.Element => {
	const gradients = React.useMemo((): Array<JSX.Element | undefined> => {
		return data.map((d, i): JSX.Element | undefined => {
			if (
				data.length > i + 1 &&
				i > 0 &&
				d.color === data[i - 1].color &&
				d.color === data[i + 1].color
			) {
				return undefined;
			}
			return (
				<stop
					offset={Number((d.distance / xMax).toFixed(10))}
					stopColor={d.color}
					stopOpacity={1}
					key={`elevationGrade-${i}`}
				/>
			);
		});
	}, [data]) as JSX.Element[];

	// if (!gradients) return <></>;

	return <>{gradients}</>;
};

export default GradeGradient;
