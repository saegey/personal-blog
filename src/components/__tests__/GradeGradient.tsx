/**
 * @jest-environment jsdom
 */
import renderer from 'react-test-renderer'

import GradeGradient from '../GradeGradient'

describe('GradGradient', () => {
  it('renders', () => {
    const json = renderer
      .create(
        <GradeGradient data={[{ distance: 5, color: 'red' }]} xMax={20} />
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })
})
