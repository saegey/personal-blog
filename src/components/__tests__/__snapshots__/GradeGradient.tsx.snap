// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradGradient renders 1`] = `
<stop
  offset={0.25}
  stopColor="red"
  stopOpacity={1}
/>
`;
